zrml/swaps/src/benchmarks.rs

```diff
@@ -395,57 +475,133 @@ benchmarks! {
             None,
             ScoringRule::CPMM,
-            false
+            false,
+            None,
         );
         let _ = Pallet::<T>::mutate_pool(pool_id, |pool| {
-            pool.pool_status = PoolStatus::Closed; Ok(())
+            pool.pool_status = PoolStatus::Closed;
+            Ok(())
         });
     }: {
-        Pallet::<T>::clean_up_pool(
-            &MarketType::Categorical(a as u16),
-            pool_id, &OutcomeReport::Categorical(0),
-            &account("ScrapCollector", 0, 0)
+        Pallet::<T>::clean_up_pool_categorical(
+            pool_id,
+            &OutcomeReport::Categorical(0),
+            &account("ScrapCollector", 0, 0),
         )?;
     }
 
     swap_exact_amount_in_cpmm {
-        let a = T::MaxAssets::get();
+        // We're trying to get as many iterations in `bpow_approx` as possible. Experiments have
+        // shown that y = 3/4, weight_ratio=1/2 (almost) maximizes the number of iterations for
+        // calculating y^r within the set of values allowed in `swap_exact_amount_in` (see
+        // `math::calc_out_given_in`). To get these values, we use the following parameters:
+        // amount_in = 1/3 * balance_in, weight_in = 1, weight_out = 2.
+        let asset_count = T::MaxAssets::get();
+        let balance: BalanceOf<T> = T::MinLiquidity::get();
+        let asset_amount_in: BalanceOf<T> = bmul(
+            balance.saturated_into(),
+            T::MaxInRatio::get().saturated_into(),
+        )
+        .unwrap()
+        .saturated_into();
+        let weight_in = T::MinWeight::get();
+        let weight_out = 2 * weight_in;
+        let mut weights = vec![weight_in; asset_count as usize];
+        weights[asset_count as usize - 1] = weight_out;
         let caller: T::AccountId = whitelisted_caller();
-        let (pool_id, assets, ..) = bench_create_pool::<T>(caller.clone(), Some(a as usize),
-            Some(T::MinLiquidity::get() * 2u32.into()), ScoringRule::CPMM, false);
-        let asset_amount_in: BalanceOf<T> = BASE.saturated_into();
-        let min_asset_amount_out: Option<BalanceOf<T>> = Some(0u32.into());
-        let max_price = Some(T::MinLiquidity::get() * 2u32.into());
-    }: swap_exact_amount_in(RawOrigin::Signed(caller), pool_id, assets[0], asset_amount_in,
-            assets[T::MaxAssets::get() as usize - 1], min_asset_amount_out, max_price)
+        let (pool_id, assets, ..) = bench_create_pool::<T>(
+            caller.clone(),
+            Some(asset_count as usize),
+            Some(balance),
+            ScoringRule::CPMM,
+            false,
+            Some(weights),
+        );
+        let asset_in = assets[0];
+        T::AssetManager::deposit(asset_in, &caller, u64::MAX.saturated_into()).unwrap();
+        let asset_out = assets[asset_count as usize - 1];
+        let min_asset_amount_out: Option<BalanceOf<T>> = Some(0u128.saturated_into());
+        let max_price = Some(u128::MAX.saturated_into());
+    }: swap_exact_amount_in(
```

Comment:
That is a little bit misleading, because there exists an extrinsic called `swap_exact_amount_in` and a same called function. Here you are calling the extrinsic, because you use the origin instead of the account id as the first parameter. In the [function `swap_exact_amount_in` you use this benchmark](https://github.com/zeitgeistpm/zeitgeist/blob/bb3e6185bc8d2cf9e849c1336ea7e52e817c71ed/zrml/swaps/src/lib.rs#L2153) result as return value. This might grow into a problem, when we add overhead in the extrinsic, that weight will automatically get added to the sub-function additionally.

Mitigation Approach: Maybe benchmark the sub-function and not the extrinsic. But because there is no overhead at the moment in the extrinsic, this is a nitpick.